README.md
# shipcat
![Shipcat logo](logo/shipcat_logo.png "Shipcat")

[![CircleCI](https://circleci.com/gh/babylonhealth/shipcat.svg?style=shield)](https://circleci.com/gh/babylonhealth/shipcat)
[![Docker Repository on Quay](https://quay.io/repository/babylonhealth/kubecat/status "Docker Repository on Quay")](https://quay.io/repository/babylonhealth/kubecat?tab=tags)

A standardisation tool and security layer on top of `kubernetes` to config manage microservices. Developers write manifests:

```yaml
name: webapp
image: clux/webapp-rs
version: 0.2.0
env:
  DATABASE_URL: IN_VAULT
resources:
  requests:
    cpu: 100m
    memory: 100Mi
  limits:
    cpu: 300m
    memory: 300Mi
replicaCount: 2
health:
  uri: /health
httpPort: 8000
regions:
- minikube
metadata:
  team: Doves
  repo: https://github.com/clux/webapp-rs
```

and `shipcat` creates a 2 replica [kubernetes deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) for [this sample webapp](https://github.com/clux/webapp-rs), with a [health check](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/) to ensure smooth upgrades. Contacts will be slack notified on upgrades.

Secrets are managed by [Vault](https://www.vaultproject.io/) and resolved by `shipcat` pre-merge, and pre-upgrade.

## Documentation
Browse the API documentation, or the setup guides available at:

- [Introduction to shipcat](https://github.com/babylonhealth/shipcat/blob/master/doc/intro.md)
- [Shipcat Definitions](https://babylonhealth.github.io/shipcat/shipcat_definitions/index.html)
- [Setup for operations](./doc/reconcile-ci.md)
- [Building](https://github.com/babylonhealth/shipcat/blob/master/doc/building.md)
- [Clusters & Regions](https://github.com/babylonhealth/shipcat/blob/master/doc/clusters.md)
- [Extending shipcat](https://github.com/babylonhealth/shipcat/blob/master/doc/extending.md)
- [Templates](https://github.com/babylonhealth/shipcat/blob/master/doc/templates.md)
- [Vault](https://github.com/babylonhealth/shipcat/blob/master/doc/vault.md)
- [Error handling](https://github.com/babylonhealth/shipcat/blob/master/doc/errors.md)
- [Nautical terminology](https://en.wikipedia.org/wiki/Ship%27s_cat)

## Components
Shipcat is made up of three main components:

- [shipcat_definitions](https://babylonhealth.github.io/shipcat/shipcat_definitions/index.html) - allowed syntax in our kube clusters - `manifest.yml` + `shipcat.conf`
- [shipcat](https://github.com/babylonhealth/shipcat/tree/master/shipcat_cli) - the pipeline cli and validator useable by developers and CI
- [raftcat](https://github.com/babylonhealth/shipcat/tree/master/raftcat) - an kubernetes api/watcher that reads the `shipcatmanifests` custom resource

## Integrations
While shipcat mainly deals with kubernetes, there are extensive and optional integrations with:

- [Vault](https://www.vaultproject.io/)
- [Kong](https://konghq.com/)
- [StatusCake](https://www.statuscake.com/)
- [Slack](https://slack.com/)

and some minor convenience integrations from common technologies like: [Grafana](https://grafana.com/), [CircleCI](https://circleci.com/), [Quay.io](https://quay.io/), [logz.io](https://logz.io/), [Sentry](https://sentry.io/), [New Relic](https://newrelic.com/)

## CLI installation

- Mac/Linux users can install from the [releases page](https://github.com/babylonhealth/shipcat/releases)
- Users with [rust](https://rustup.rs/) installed can use `git pull && cargo build`
- Babylon employees can `brew install shipcat` or `brew update && brew upgrade shipcat` via the internal brew tap

See the [building guide](https://github.com/babylonhealth/shipcat/blob/master/doc/building.md), for setting up auto-complete, and being able to use from outside a manifests repo.

## CLI Usage
Define your `manifest.yml` file in a [manifests repo](https://github.com/babylonhealth/shipcat/blob/master/examples), make sure `shipcat validate` passes.

You either need to have a `~/.kube/config` whose `current-context` is set to the shipcat region you wish to validate, or pass the shipcat region in explicitly with `-r region`.

If you have `vault` read credentials (a `VAULT_TOKEN` evar, or a `~/.vault-token` file) you can validate secret existence and generate the completed manifest (values):

```sh
shipcat validate webapp --secrets

# Generate completed manifest (what's passed to your chart)
shipcat values webapp -s
```

If you have `helm` installed you can generate the helm template via the associated helm chart:

```sh
# Pass completed manifest to helm template
shipcat template webapp
```

## License
Apache 2.0 licensed. See LICENSE for details.


CHANGELOG.md
0.151.2 / 2020-04-08
====================
  * Minified `shipcat diff` via reconcile now always hides secret objects
  * Statefulset rollout tracking now returns AFTER rollout is complete - #447

0.151.1 / 2020-04-06
====================
  * fix `delete` not being able to serialize responses in reconcile - #437

0.151.0 / 2020-04-02
====================
  * fix a bug in `shipcat get codeowners` coercing a user.github to null string
  * add `securityContext` as top level syntax for statefulsets - #443

0.150.0 / 2020-03-19
====================
  * `manifest.autoScaling` upgraded from hpa v2beta1 to hpa v2beta2 - #255
  * `manifest.autoScaling` now supports all `HorizontalPodAutoscaler` features
  * Fix bug in tracking module during rollbacks - #435

0.149.0 / 2020-03-18
====================
  * upgrade tracking now uses kube for api tracking
  * kubectl module majorly trimmed down
  * new minial cmd `shipcat cluster crd install` to configure a `Region`
  * `shipcat status` uses `kube` and gives better info
  * `shipcat apply` has a progress bar using `indicatif`
  * crd generation now done via `kube-derive` (crds.rs trimmed)

0.148.4 / 2020-03-18
====================
  * add python "Schema" library to kubecat docker image

0.148.3 / 2020-03-02
====================
  * `shipcat cluster check` now useable offline
    - adds comma separated `--skip-kinds` flag for less restrictive label checks

0.148.0 / 2020-02-26
====================
  * TLS transport now uses rustls rather than openssl

0.147.0 / 2020-02-26
====================
  * Convert shipcat to use async/await internally
    - tokio "fs" feature in shipcat_filebacked
    - tokio "process" feature in shipcat_cli
    - kube 0.25.0 for k8s api calls
    * Remove rayon + threadpool dependencies
    - cluster + validate/verify commands now just use tokio "stream" feature
  * Upgrade reqwest to 0.10 everywhere in shipcat cli (-15 deps)
  * Upgrade url to 2.X and remove url_serde dependency
  * Added optional `metadata.context` field
  * Allow empty `producers` +  `consumers` in `eventStreams`

0.146.0 / 2020-02-10
====================
  * `Manifest.eventStreams` syntax added for Kafka topic and Zookeepr ACL creation - #381

0.145.0 / 2020-02-07
====================
  * service deletion now gets a slack warning
  * `shipcat restart` now also restarts worker deployments
  * `shipcatmanifest` crd status object now contains a `lastSuccessfulRolloutVersion`

0.144.1 / 2020-02-05
====================
  * shipcat diff `-k` or `--current` removed (now default)
  * new `--mock` introduced in `shipcat diff` to get old behaviour
  * `workload` new property on `Manifest` (to support Deployment/Statefulset switches)

0.143.0 / 2020-01-31
====================
  * Tiller support removed
  * `shipcat template -c MYSVC` properly validates template labels
  * `shipcat cluster check` is a region-wide variant of `shipcat template -c`

0.142.0 / 2020-01-07
====================
  * ReconciliationMode::CrdOwned new default
  * Tiller support fully deprecated and will be removed soon

0.134.1 / 2019-10-23
====================
  * Speculative `shipcat self-upgrade`
  * Can be invoked automatically with `SHIPCAT_AUTOUPGRADE=1` evar.
  * Can avoid rate limiting with `SHIPCAT_AUTOUPGRADE_TOKEN` to github token with `repo:read` scope.

0.133.0 / 2019-10-18
====================
  * introductory syntax for `newRelic` and `sentry`
  * kong: supporting multiple kong apis per service
  * `jobs` syntax removed
  * version warning now always present, but ignored unless --strict-version-check set
  * `shipcat restart {service}` subcommand added
  * `shipcat diff` improvements
  * `shipcat port-forward` improvements
  * better error messages for missing executable dependencies

0.132.0 / 2019-10-08
====================
  * `.rbac` key in manifests now support arbitrary verbs and nouns
  * better error messages for `teams.yml` errors
  * readiness / liveness probes support timeout properly

0.131.0 / 2019-10-07
====================
  * kong: json cookie header deprecated parameters removed

0.130.0 / 2019-10-01
====================
  * `shipcat` accepts team owners in `teams.yml`
  * `shipcat.conf` will phase out `teams` vector from this release

0.129.0 / 2019-09-26
====================
  * `diff` can can compare regions
  * `apply` now resilient against manifest schema changes (bug)

0.127.1 / 2019-09-16
====================
  * Diff now correctly detects kubectl diff output and filters out generation

0.127.0 / 2019-09-13
====================
  * First release to support upgrades withut tiller.
  * `ReconciliationMode::CrdOwned` and can be set per-region in `shipcat.conf`.
  * Tested properly in examples/ directory
  * Charts must provide the following properties in metadata now:

```yaml
metadata:
  labels:
    app.kubernetes.io/name: {{ .Values.name }}
    app.kubernetes.io/version: {{ .Values.version }}
    app.kubernetes.io/managed-by: shipcat
  ownerReferences:
  - apiVersion: babylontech.co.uk/v1
    kind: ShipcatManifest
    controller: false
    name: {{ .Values.name }}
    uid: {{ .Values.uid }}
```

 * `shipcat diff` rewritten, uses kubectl by default
   - `--minify` or `-m` added (to minify the diff)
   - `--obfuscate` added (to hide secrets)
   - `--current` or `-k` added (to use actual uids/versions from kube)
 * `shipcat template` improved
   - `--check` or `-c` added (verifies chart assumptions)
   - `--current` or `-k` added (to use actual uids/versions from kube)

0.126.0 / 2019-09-03
====================
  * `reconciliationMode` default is now `CrdStatus` rather than `CrdVersioned`
  * Kubernetes default server requirement now 1.12


src/kube.rs
use super::{Result, Manifest};
use regex::Regex;
use serde_yaml;
use chrono::{Utc, DateTime};

fn kexec(args: Vec<String>) -> Result<()> {
    use std::process::Command;
    debug!("kubectl {}", args.join(" "));
    let s = Command::new("kubectl").args(&args).status()?;
    if !s.success() {
        bail!("Subprocess failure from kubectl: {}", s.code().unwrap_or(1001))
    }
    Ok(())
}
fn kout(args: Vec<String>) -> Result<String> {
    use std::process::Command;
    debug!("kubectl {}", args.join(" "));
    let s = Command::new("kubectl").args(&args).output()?;
    let out : String = String::from_utf8_lossy(&s.stdout).into();
    let err : String = String::from_utf8_lossy(&s.stderr).into();
    if !err.is_empty() {
        warn!("kubectl {} stderr: {}", args.join(" "), err.trim());
    }
    // kubectl keeps returning opening and closing apostrophes - strip them:
    if out.len() > 2 && out.chars().next() == Some('\'') {
        let res = out.split('\'').collect::<Vec<_>>()[1];
        return Ok(res.trim().into());
    }
    Ok(out)
}

pub fn current_context() -> Result<String> {
    let mut res = kout(vec!["config".into(), "current-context".into()]).map_err(|e| {
        error!("Failed to Get kubectl config current-context. Is kubectl installed?");
        e
    })?;
    let len = res.len();
    if res.ends_with('\n') {
        res.truncate(len - 1);
    }
    Ok(res)
}

fn rollout_status(mf: &Manifest) -> Result<bool> {
    // TODO: handle more than one deployment
    // Even if this were called 10 times with 1/10th of waiting time, we still can't wait:
    // - we'd still need to check other deploys...
    // - we'd have to deal with `kubectl rollout status` not having a timeout flag..
    let statusvec = vec![
        "rollout".into(),
        "status".into(),
        format!("deployment/{}", mf.name.clone()), // always one deployment with same name
        format!("-n={}", mf.namespace),
        "--watch=false".into(), // always just print current status
    ];
    let rollres = kout(statusvec)?;
    debug!("{}", rollres);
    if rollres.contains("successfully rolled out") {
        Ok(true)
    } else {
        // TODO: check if any of the new pods have restarts in them
        // will avoid waiting for the full time
        Ok(false)
    }
}

/// A replacement for helm upgrade's --wait and --timeout
pub fn await_rollout_status(mf: &Manifest) -> Result<bool> {
    use std::{thread, time};
    // Check for rollout progress
    let waittime = mf.estimate_wait_time();
    let sec = time::Duration::from_millis(1000);
    // if this is called immediately after apply/upgrade, resources might not exist yet
    match rollout_status(&mf) {
        Ok(true) => return Ok(true), // can also insta-succeed on "noops"
        Ok(false) => debug!("Ignoring rollout failure right after upgrade"),
        Err(e) => warn!("Ignoring rollout failure right after upgrade: {}", e),
    };
    info!("Waiting {}s for deployment {} to rollout (not ready yet)", waittime, mf.name);
    for i in 1..10 {
        trace!("poll iteration {}", i);
        let mut waited = 0;
        // sleep until 1/10th of estimated upgrade time and poll for status
        while waited < waittime/10 {
            waited += 1;
            trace!("sleep 1s (waited {})", waited);
            thread::sleep(sec);
        }
        if rollout_status(&mf)? {
            return Ok(true)
        }
    }
    Ok(false) // timeout
}

fn get_pods(mf: &Manifest) -> Result<String> {
    //kubectl get pods -l=app=$* -o jsonpath='{.items[*].metadata.name}'
    let podargs = vec![
        "get".into(),
        "pods".into(),
        format!("-l=app={}", mf.name),
        format!("-n={}", mf.namespace),
        "-o".into(),
        "jsonpath='{.items[*].metadata.name}'".into(),
    ];
    // TODO: filter out ones not running conditionally - exec wont work with this
    let podsres = kout(podargs)?;
    debug!("Active pods: {:?}", podsres);
    Ok(podsres)
}

/// Return non-running or partially ready pods
fn get_broken_pods(mf: &Manifest) -> Result<(String, Vec<String>)> {
    let podargs = vec![
        "get".into(),
        "pods".into(),
        format!("-l=app={}", mf.name),
        format!("-n={}", mf.namespace),
        format!("--no-headers"),
    ];
    let podres = kout(podargs)?;
    let mut bpods = vec![];
    let status_re = Regex::new(r" (?P<ready>\d+)/(?P<total>\d+) ").unwrap();
    for l in podres.lines() {
        if !l.contains("Running") {
            if let Some(p) = l.split(' ').next() {
                warn!("Found pod not running: {}", p);
                bpods.push(p.into());
            }
        }
        else if let Some(caps) = status_re.captures(l) {
            if &caps["ready"] != &caps["total"] {
                if let Some(p) = l.split(' ').next() {
                    warn!("Found pod with less than necessary containers healthy: {}", p);
                    bpods.push(p.into());
                }
            }
        }
    }
    Ok((podres, bpods))
}

/// Debug helper when upgrades fail
///
/// Prints log excerpts and events for broken pods.
/// Typically enough to figure out why upgrades broke.
pub fn debug(mf: &Manifest) -> Result<()> {
    let (podres, pods) = get_broken_pods(&mf)?;
    if pods.is_empty() {
        info!("No broken pods found");
        info!("Pod statuses:\n{}", podres);
    } else {
        warn!("Pod statuses:\n{}", podres);
    }
    for pod in pods.clone() {
        warn!("Debugging non-running pod {}", pod);
        let logvec = vec![
            "logs".into(),
            pod.clone(),
            mf.name.clone(),
            format!("-n={}", mf.namespace),
            format!("--tail=30").into(),
        ];
        match kout(logvec) {
            Ok(l) => {
                if l == "" {
                    warn!("No logs for pod {} found", pod);
                } else {
                    warn!("Last 30 log lines:");
                    print!("{}\n", l);
                }
            },
            Err(e) => {
                warn!("Failed to get logs from {}: {}", pod, e)
            }
        }
    }

    for pod in pods {
        warn!("Describing events for pod {}", pod);
        let descvec = vec![
            "describe".into(),
            "pod".into(),
            pod.clone(),
            format!("-n={}", mf.namespace),
        ];

        match kout(descvec) {
            Ok(mut o) => {
                if let Some(idx) = o.find("Events:\n") {
                    print!("{}\n", o.split_off(idx))
                }
                else {
                    // Not printing in this case, tons of secrets in here
                    warn!("Unable to find events for pod {}", pod);
                }
            },
            Err(e) => {
                warn!("Failed to describe {}: {}", pod, e)
            }
        }
    }
    // ignore errors from here atm - it's mostly here as a best effort helper
    let _ = debug_active_replicasets(mf);
    Ok(())
}


// Parsed replica status from a deployment description
#[derive(Clone, Debug)]
struct ReplicaStatus {
    /// Name of ReplicaSet
    pub name: String,
    /// Total replicas expected
    pub total: u32,
    /// Total replicas running
    pub running: u32,
}

// limited parsed struct from kube ReplicaSet yaml
#[derive(Deserialize)]
struct ReplicaSetVal {
    // rollout status of the RS
    status: ReplicaInfoVal,
    // metadata for timestamps
    metadata: ReplicaMetadataVal,
}
#[derive(Deserialize, Debug)]
struct ReplicaMetadataVal {
    creationTimestamp: DateTime<Utc>, // e.g. 2018-05-17T10:08:37Z
}
#[derive(Deserialize, Debug)]
struct ReplicaInfoVal {
    // `observedGeneration` <- not unique across replicasets so useless
    // Currently available replicas (ready for minReadySeconds) from this generation
    #[serde(default)] // 0 if missing
    availableReplicas: u32,
    /// Currently ready replilas from this generation (weaker than above)
    #[serde(default)] // 0 if missing
    readyReplicas: u32,
    /// Most recently oberved number of replicas
    replicas: u32,
}


/// Simplified ReplicaSet struct
///
/// Created by combining deployment description with appropriate replicaset yamls
#[derive(Debug)]
pub struct ReplicaSet {
    /// Name of replicaset
    pub name: String,
    /// Available replicas (has been ready for minReadySeconds)
    pub available: u32,
    /// Total replicas in set
    pub total: u32,
    /// Created timestamp (used for ordering)
    pub created: DateTime<Utc>,
}

// Finds affected replicatsets, and checks the state of them.
fn find_active_replicasets(mf: &Manifest) -> Result<Vec<ReplicaSet>> {
    // find relevant replicasets
    // NB: not returned via `k get deploy {name} -oyaml` - have to scrape describe..
    let descvec = vec![
        "describe".into(),
        "deploy".into(),
        mf.name.clone(),
        format!("-n={}", mf.namespace),
    ];
    // Finding the affected replicasets:
    let rs_re = Regex::new(r"(Old|New)ReplicaSets?:\s+(?P<rs>\S+)\s+\((?P<running>\d+)/(?P<total>\d+)").unwrap();
    let deployres = kout(descvec)?;
    let mut sets = vec![];
    for l in deployres.lines() {
        if let Some(caps) = rs_re.captures(l) {
            sets.push(ReplicaStatus {
                name: caps["rs"].to_string(),
                running: caps["running"].parse()?,
                total: caps["total"].parse()?,
            });
        }
    }
    // Query each of the affected replicasets for info using yaml api:
    let mut completesets = vec![];
    for rs in sets {
        // Find total counts from each replicaset
        let getvec = vec![
            "get".into(),
            "rs".into(),
            rs.name.clone(),
            format!("-n={}", mf.namespace),
            "-oyaml".into(),
        ];
        let getres = kout(getvec)?;
        let rv : ReplicaSetVal = serde_yaml::from_str(&getres)?;
        let ri = rv.status;
        let res = ReplicaSet {
            name: rs.name,
            available: ri.availableReplicas,
            total: ri.replicas,
            created: rv.metadata.creationTimestamp
        };
        completesets.push(res);
    }
    Ok(completesets)
}

// Debug status of active replicasets post-upgrade helpful info
fn debug_active_replicasets(mf: &Manifest) -> Result<()> {
    let sets = find_active_replicasets(mf)?;
    if sets.len() > 1 {
        warn!("ReplicaSets: {:?}", sets);
    }
    if let Some(latest) = sets.iter().max_by_key(|x| x.created.timestamp()) {
        info!("Latest {:?}", latest);
        if latest.available > 0 && latest.available < latest.total {
            warn!("Some replicas successfully rolled out - maybe a higher timeout would help?");
        }
        else if latest.available == 0{
            warn!("No replicas were rolled out fast enough ({} secs)", mf.estimate_wait_time());
            warn!("Your application might be crashing, or fail to respond to healthchecks in time");
            warn!("Current health check is set to {:?}", mf.health);
        }
    } else {
        warn!("No active replicasets found");
    }
    Ok(())
}

/// Print upgrade status of current replicaset rollout
pub fn debug_rollout_status(mf: &Manifest) -> Result<()> {
    let mut sets = find_active_replicasets(mf)?;
    if sets.len() == 2 {
        sets.sort_unstable_by(|x,y| x.created.timestamp().cmp(&y.created.timestamp()));
        let old = sets.first().unwrap();
        let new = sets.last().unwrap();
        info!("{} upgrade status: old {}/{} -  new {}/{} ", mf.name,
            old.available, old.total,
            new.available, new.total
        );
    }
    Ok(())
}


/// Shell into all pods associated with a service
///
/// Optionally specify the arbitrary pod index from kubectl get pods
pub fn shell(mf: &Manifest, desiredpod: Option<usize>, cmd: Option<Vec<&str>>) -> Result<()> {
    // TODO: kubectl auth can-i create pods/exec
    let podsres = get_pods(&mf)?;
    let pods = podsres.split(' ').collect::<Vec<_>>();
    let pnr = desiredpod.unwrap_or(0);
    if let Some(p) = pods.get(pnr) {
        debug!("Shelling into {}", p);
        //kubectl exec -it $pod sh
        let mut execargs = vec![
            "exec".into(),
            format!("-n={}", mf.namespace),
            "-it".into(),
            p.to_string(),
        ];
        if let Some(cmdu) = cmd.clone() {
            for c in cmdu {
                execargs.push(c.into())
            }
        } else {
            let trybash = vec![
                "exec".into(),
                format!("-n={}", mf.namespace),
                p.to_string(),
                "which".into(),
                "bash".into(),
            ];
            // kubectl exec $pod which bash
            // returns a non-zero rc if not found generally
              let shexe = match kexec(trybash) {
                Ok(o) => {
                    debug!("Got {:?}", o);
                    "bash".into()
                },
                Err(e) => {
                    warn!("No bash in container, falling back to `sh`");
                    debug!("Error: {}", e);
                    "sh".into()
                }
            };
            execargs.push(shexe);
        }
        kexec(execargs)?;
    } else {
        bail!("Pod {} not found for service {}", pnr, &mf.name);
    }
    Ok(())
}


/// Port forward a port to localhost
pub fn port_forward(mf: &Manifest, desiredpod: Option<usize>) -> Result<()> {
    // TODO: kubectl auth can-i create something?
    let podsres = get_pods(&mf)?;
    let pods = podsres.split(' ').collect::<Vec<_>>();
    let pnr = desiredpod.unwrap_or(0);
    let port = mf.httpPort.unwrap();
    // first 1024 ports need sudo so avoid that
    let localport = if port <= 1024 { 7777 } else { port };
    if let Some(p) = pods.get(pnr) {
        debug!("Port forwarding kube pod {} to localhost:{}", p, localport);
        //kubectl port-forward $pod localport:httpPort
        let mut pfargs = vec![
            format!("-n={}", mf.namespace),
            "port-forward".into(),
            p.to_string(),
            format!("{}:{}", port, port)
        ];
        kexec(pfargs)?;
    } else {
        bail!("Pod {} not found for service {}", pnr, &mf.name);
    }
    Ok(())
}


#[cfg(test)]
mod tests {
    use dirs;
    use super::current_context;

    #[test]
    fn validate_ctx() {
        let kubecfg = dirs::home_dir().unwrap().join(".kube").join("config");
        if kubecfg.is_file() {
            let ctx = current_context().unwrap();
            assert_eq!(ctx, "dev-uk".to_string());
        }
    }
}
